lib/Ebook.php

```diff
@@ -306,272 +801,363 @@ public function __construct(?string $wwwFilesystemPath = null){
+		$ebookFromFilesystem->WikipediaUrl = Ebook::NullIfEmpty($xml->xpath('/package/metadata/meta[@property="se:url.encyclopedia.wikipedia"][not(@refines)]'));
 
 		// Next the page scan source URLs.
+		$sources = [];
 		foreach($xml->xpath('/package/metadata/dc:source') ?: [] as $element){
 			$e = (string)$element;
 			if(mb_stripos($e, 'gutenberg.org/') !== false){
-				$this->Sources[] = new EbookSource(EbookSourceType::ProjectGutenberg, $e);
+				$sources[] = EbookSource::FromTypeAndUrl(EbookSourceType::ProjectGutenberg, $e);
 			}
 			elseif(mb_stripos($e, 'gutenberg.net.au/') !== false){
-				$this->Sources[] = new EbookSource(EbookSourceType::ProjectGutenbergAustralia, $e);
+				$sources[] = EbookSource::FromTypeAndUrl(EbookSourceType::ProjectGutenbergAustralia, $e);
 			}
 			elseif(mb_stripos($e, 'gutenberg.ca/') !== false){
-				$this->Sources[] = new EbookSource(EbookSourceType::ProjectGutenbergCanada, $e);
+				$sources[] = EbookSource::FromTypeAndUrl(EbookSourceType::ProjectGutenbergCanada, $e);
 			}
 			elseif(mb_stripos($e, 'archive.org/details') !== false){
 				// `/details` excludes Wayback Machine URLs which may sometimes occur, for example in Lyrical Ballads
-				$this->Sources[] = new EbookSource(EbookSourceType::InternetArchive, $e);
+				$sources[] = EbookSource::FromTypeAndUrl(EbookSourceType::InternetArchive, $e);
 			}
 			elseif(mb_stripos($e, 'hathitrust.org/') !== false){
-				$this->Sources[] = new EbookSource(EbookSourceType::HathiTrust, $e);
+				$sources[] = EbookSource::FromTypeAndUrl(EbookSourceType::HathiTrust, $e);
 			}
 			elseif(mb_stripos($e, 'wikisource.org/') !== false){
-				$this->Sources[] = new EbookSource(EbookSourceType::Wikisource, $e);
+				$sources[] = EbookSource::FromTypeAndUrl(EbookSourceType::Wikisource, $e);
 			}
 			elseif(mb_stripos($e, 'books.google.com/') !== false || mb_stripos($e, 'google.com/books/') !== false){
-				$this->Sources[] = new EbookSource(EbookSourceType::GoogleBooks, $e);
+				$sources[] = EbookSource::FromTypeAndUrl(EbookSourceType::GoogleBooks, $e);
 			}
 			elseif(mb_stripos($e, 'www.fadedpage.com') !== false){
-				$this->Sources[] = new EbookSource(EbookSourceType::FadedPage, $e);
+				$sources[] = EbookSource::FromTypeAndUrl(EbookSourceType::FadedPage, $e);
 			}
 			else{
-				$this->Sources[] = new EbookSource(EbookSourceType::Other, $e);
+				$sources[] = EbookSource::FromTypeAndUrl(EbookSourceType::Other, $e);
 			}
 		}
+		$ebookFromFilesystem->Sources = $sources;
 
 		// Next the GitHub URLs.
-		$this->GitHubUrl = $this->NullIfEmpty($xml->xpath('/package/metadata/meta[@property="se:url.vcs.github"][not(@refines)]'));
+		$ebookFromFilesystem->GitHubUrl = Ebook::NullIfEmpty($xml->xpath('/package/metadata/meta[@property="se:url.vcs.github"][not(@refines)]'));
+
+		return $ebookFromFilesystem;
+	}
+
+
+	// *******
+	// METHODS
+	// *******
+
+	/**
+	 * @throws Exceptions\ValidationException
+	 */
+	public function Validate(): void{
```

Comment:
Here we must check that the properties you compare are `isset()` first .If they aren't set, then testing against `''` will be a fatal error. You should also trim everything here as well.

Reply:
Could you clarify something about these properties? It seems there are some that are truly required and marked `NOT NULL` in the DB schema:

* `Identifier`
* `WwwFilesystemPath`
* `RepoFilesystemPath`
* `Title`
* `Description`
* `LongDescription`
* `WordCount`
* `ReadingEase`
* `EbookCreated`
* `EbookUpdated`
* `TextSinglePageByteCount`
* `IndexableText`

For those properties, I trimmed the string ones and checked the others with `isset()`.

Then there are properties that don't seem strictly required and are allowed to be `NULL` in the DB schema:

* `KindleCoverUrl`
* `EpubUrl`
* `AdvancedEpubUrl`
* `KepubUrl`
* `Azw3Url`
* `DistCoverUrl`
* `FullTitle`
* `AlternateTitle`
* `GitHubUrl`
* `WikipediaUrl`

For those, I trimmed them if they are not `null`, but I didn't coerce them from `null` to `''`. That seems right to me.  Some of those URLs don't seem optional to me, though, but maybe formats change over time.

One place where I changed my mind was the `Language` property. It's currently allowed to be `NULL` in the DB, but that seems wrong. I changed the code to not allow a `null` in `$this->Language`, and if you agree I'll update the DB schema, too.

In my 1000+ ebooks, all of them have `Language` set:

```
> select * from Ebooks where Language IS NULL\G
```

Reply:
In general, if it's `NOT NULL` in the DB then the property *must* have a value during validation.

Also in general, all string properties should be trimmed during validation.

For properties that *are* nullable, if it's a string, then an empty string should be changed to `null` before saving. This example covers all cases:

````php
function Validate(){
    $this->NullableStringProperty = trim($this->NullableStringProperty ?? '');
    if($this->NullableStringProperty == ''){
        $this->NullableStringProperty = null;
    }

    if(!isset($this->NotNullableProperty)){
        // Validation error - property that is not nullable has not been set before validation.
    }
}
````

Does that answer your question?

Reply:
> Does that answer your question?

Yep, thanks a bunch for the example code. I that approach to the `Validate()` methods in `Ebook`, `Collection`, `Contributor`, `EbookSource`, `EbookTag`, `GitCommit`, and `LocSubject`. Regarding the "Possible validation error?" for non-nullable string properties, I treated them as validation errors with the same exception as if they property was `null`. That seemed like what we wanted.